### PR TITLE
Change unchecked unwrapping to unwrap_or_default for Args-trait (#1845)

### PR DESCRIPTION
**Pull request by ChrisRega**
Simply changing the unwrap to unwrap_or_default in order to keep the number of arguments identical to the original iterator contents.

### CODE REVIEW

**Comment by sylvestre** [comment]
Could you please add a test verifying that the old issue doesn't happen anymore?
thanks

**Comment by ChrisRega** [comment]
> Could you please add a test verifying that the old issue doesn't happen anymore?
> thanks

Added unit-testing for the desired behaviour.

**Comment by Arcterus** [comment]
I’m not sure this is ideal, given that it silently ignores non-UTF-8 arguments.

**Comment by ChrisRega** [comment]
> I’m not sure this is ideal, given that it silently ignores non-UTF-8 arguments.

We could either go for some eprintln-warning that an argument was ignored or make this function return a result itself and make the client-code deal with it. There are around 50 usages of it, most look like they collect the arguments before giving it to cl argument parsing. Dealing with results there would produce a lot of very similar code, which then has to be maintained. I see two possible solutions:

1. Ignore the arguments that have broken encoding (with or without warning) and try to continue the program
2. Abort execution alltogether since the user probably wanted to do something with this argument and we clearly can't interpret his intention.

Now that I think a bit more about it, option 2 seems even more sane than just ignoring a parameter (even with a warning).

**Comment by sylvestre** [comment]
Would it be possible to add tests at the binary leve? For example in https://github.com/uutils/coreutils/blob/master/tests/by-util/test_basename.rs ?
thanks

**Comment by ChrisRega** [comment]
> Would it be possible to add tests at the binary leve? For example in https://github.com/uutils/coreutils/blob/master/tests/by-util/test_basename.rs ?
> thanks

Absolutely, will implement some tomorrow! Thanks for all the feedback

**Comment by ChrisRega** [comment]
> > Would it be possible to add tests at the binary leve? For example in https://github.com/uutils/coreutils/blob/master/tests/by-util/test_basename.rs ?
> > thanks
> 
> Absolutely, will implement some tomorrow! Thanks for all the feedback

Turns out, that implementing a test for basename is not trivial after all. UCommand parses its arguments with unwrap directly without ever calling collect_str.
The next problem is with the result of the test, basename with a broken argument should fail like with no argument. It does not, so i thought about changing the behaviour of collect_str to drop items with broken encoding. But this causes more problems in other tests. I will try to investigate this issue a bit more in the next days. I am not satisfied with the current behaviour.

**Review by sylvestre — CHANGES_REQUESTED** [review]
OK, please let me know when it is ready to be reviewed!

File: src/uucore/src/lib/lib.rs [hunk 1]
```diff
@@ -67,10 +67,63 @@ pub use crate::features::wide;
...
+        match full_conversion {
+            true => ConversionResult::FullyDecoded(result_vector),
+            false => ConversionResult::Lossy(result_vector),
+        }
+    }
+
+    /// convience function for a more slim interface
+    fn collect_str_lossy(self) -> ConversionResult {
+        collect_str(InvalidEncodingHandling::ConvertLossy)
     }
```
Comment (ChrisRega) [inline src/uucore/src/lib/lib.rs:127]: @sylvestre what do you think of this? It allows a slim interface with convenience functions like collect_str_lossy while still giving the callers control of what should happen in cases of error. Also, as a caller, you will get a result-like type which allows you to react. There are two example-usages: head and dirname. If you like the design, i will finish this by adding testing and changing all callers to the new API (i would figure the desired behaviour of each command to match the original coreutils).

**Comment by sylvestre** [comment]
@Arcterus what do you think about this change? Thanks

**Comment by ChrisRega** [comment]
> @Arcterus what do you think about this change? Thanks

I just added unit testing for the conversion. Feedback on this would be appreciated :)

**Comment by sylvestre** [comment]
There are some conflicts :)

Besides that, I am fine but I would like @Arcterus to review it

**Comment by ChrisRega** [comment]
> There are some conflicts :)
> 
> Besides that, I am fine but I would like @Arcterus to review it

Merged master, fixed the conflicts and updated the pr :)

**Comment by sylvestre** [comment]
Looks like it is conflicting again, sorry :/

**Comment by ChrisRega** [comment]
> Looks like it is conflicting again, sorry :/

Merged master in again and fixed some hiccups

**Comment by sylvestre** [comment]
Sorry but you still have build failures:
```

error[E0433]: failed to resolve: use of undeclared type or module `OsString`
Error:    --> D:\a\coreutils\coreutils\tests\by-util\test_basename.rs:115:21
    |
115 |       let os_string = OsString::from_wide(&source[..]);
    |                       ^^^^^^^^
    |
help: a struct with a similar name exists
    |
115 |     let os_string = String::from_wide(&source[..]);
    |                     ^^^^^^
help: consider importing one of these items
    |
1   | use crate::test_sync::fs::OsString;
    |
1   | use std::ffi::OsString;
```

File: tests/by-util/test_basename.rs [hunk 2]
```diff
@@ -84,3 +85,23 @@ fn test_no_args() {
 fn test_too_many_args() {
     expect_error(vec!["a", "b", "c"]);
 }
+
+fn test_invalid_utf8_args(os_str: &OsStr) {
```
Comment (sylvestre) [inline tests/by-util/test_basename.rs:89]: Missing
```suggestion
#[test] 
fn test_invalid_utf8_args(os_str: &OsStr) {
```

File: src/uucore/src/lib/lib.rs [hunk 3]
```diff
@@ -84,3 +164,79 @@ pub fn args() -> impl Iterator<Item = String> {
...
+            );
+        }
+        //lossy conversion string for illegal utf8
+        assert_eq!(
+            *collected_to_str.get(2).unwrap(),
+            os_str.to_os_string().to_string_lossy()
+        );
+    }
+
+    fn test_invalid_utf8_args_ignore(os_str: &OsStr) {
```
Comment (sylvestre) [inline src/uucore/src/lib/lib.rs:208]: #test missing ?

File: src/uu/unlink/src/unlink.rs [hunk 4]
```diff
@@ -17,12 +17,13 @@ use libc::{lstat, stat, unlink};
...
 use std::ffi::CString;
 use std::io::{Error, ErrorKind};
+use uucore::InvalidEncodingHandling;
 
 static NAME: &str = "unlink";
 static VERSION: &str = env!("CARGO_PKG_VERSION");
 
 pub fn uumain(args: impl uucore::Args) -> i32 {
-    let args = args.collect_str();
+    let args = args.collect_str(InvalidEncodingHandling::ConvertLossy).accept_any();
```
Comment (sylvestre) [inline src/uu/unlink/src/unlink.rs:26]: rustfmt isn't happy here :)

**Review by sylvestre — CHANGES_REQUESTED** [review]
some nits
but I would appreciate also a more detailled description  of the why (less technical)

File: src/uucore/src/lib/lib.rs (hunk 3 above)
Comment (ChrisRega) [reply src/uucore/src/lib/lib.rs:208]: See above comment

File: src/uu/unlink/src/unlink.rs (hunk 4 above)
Comment (ChrisRega) [reply src/uu/unlink/src/unlink.rs:26]: That's bad, must have gone wrong with the conflict resolution - i will fix it asap.

Comment (sylvestre) [reply src/uu/unlink/src/unlink.rs]: np

File: tests/by-util/test_basename.rs (hunk 2 above)
Comment (sylvestre) [inline tests/by-util/test_basename.rs:89]: Missing #[test] ?

File: src/uucore/src/lib/lib.rs [hunk 5]
```diff
@@ -84,3 +168,83 @@ pub fn args() -> impl Iterator<Item = String> {
...
+            );
+        }
+        //lossy conversion for string with illegal encoding is done
+        assert_eq!(
+            *collected_to_str.get(2).unwrap(),
+            os_str.to_os_string().to_string_lossy()
+        );
+    }
+
+    fn test_invalid_utf8_args_ignore(os_str: &OsStr) {
```
Comment (sylvestre) [inline src/uucore/src/lib/lib.rs:212]: ditto?

File: tests/by-util/test_basename.rs (hunk 2 above)
Comment (ChrisRega) [reply tests/by-util/test_basename.rs:89]: called by line 242: 
`#[cfg(any(unix, target_os = "redox"))]
    #[test]
    fn invalid_utf8_args_unix() {
        use std::os::unix::ffi::OsStrExt;

        let source = [0x66, 0x6f, 0x80, 0x6f];
        let os_str = OsStr::from_bytes(&source[..]);
        test_invalid_utf8_args_lossy(os_str);
        test_invalid_utf8_args_ignore(os_str);
    }`
Allows for no code duplication with illegal windows wide-string encoding (not implemented yet due to missing windows machine)

File: src/uucore/src/lib/lib.rs (hunk 5 above)
Comment (ChrisRega) [reply src/uucore/src/lib/lib.rs:212]: Same here: Called from line 248 in lib.rs allows for testing broken encoding on different platforms.

File: tests/by-util/test_basename.rs (hunk 2 above)
Comment (ChrisRega) [reply tests/by-util/test_basename.rs:89]: @sylvestre

**Pull request merged**